libraries/AP_Scripting/AP_Scripting_helpers.cpp

```diff
@@ -195,4 +195,316 @@ bool Parameter::set_default(float value)
+    }
+    cleanup = this;
+}
+
+DroneCAN_Handle::Subscriber::~Subscriber(void)
+{
+    WITH_SEMAPHORE(sem[index]);
+    unlink();
+    Payload payload;
+    while (payloads.pop(payload)) {
+        free(payload.data);
+    }
+
+    // remove from cleanup list
+    if (cleanup_next) {
+        cleanup_next->cleanup_prev = cleanup_prev;
+    }
+    if (cleanup_prev) {
+        cleanup_prev->cleanup_next = cleanup_next;
+    }
+    if (this == cleanup) {
+        cleanup = cleanup_next;
+    }
+}
+
+/*
+  handle an incoming subscribed message
+ */
+bool DroneCAN_Handle::Subscriber::handle_message(const CanardRxTransfer& transfer)
+{
+    if (transfer.transfer_type == CanardTransferTypeResponse &&
+        (node_id != transfer.source_node_id ||
+         ((transfer.transfer_id+1)&0xFF) != handle->transfer_id)) {
+        // not from right node, or not right transfer ID
+        return false;
+    }
+
+    if (payloads.space() == 0) {
+        // already have MAX_PAYLOADS messages pending, discard this one,
+        // but return true to say it was for us
+        return true;
+    }
+    Payload payload {};
+    payload.data = (uint8_t *)malloc(transfer.payload_len);
+    if (payload.data == nullptr) {
+        // discard messages if we can't allocate
+        // but return true to say it was for us
+        return true;
+    }
+
+    payload.length = transfer.payload_len;
+    payload.node_id = transfer.source_node_id;
+    payload.timestamp = transfer.timestamp_usec;
+#if CANARD_ENABLE_CANFD
+    payload.canfd = transfer.canfd;
+#endif
+
+    uint32_t bits_remaining = transfer.payload_len * 8;
+    uint32_t ofs = 0;
+
+    // decode in bytes, we don't have a canardDecodeBuffer function
+    while (bits_remaining > 0) {
+        uint8_t nbits = MIN(8U, bits_remaining);
+        canardDecodeScalar(&transfer, ofs*8, nbits, false, (void*)&payload.data[ofs]);
+        ofs++;
+        bits_remaining -= nbits;
+    }
+
+    // push to queue
+    payloads.push(payload);
+
+    return true;
+}
+
+/*
+  check for an incoming message
+ */
+int DroneCAN_Handle::check_message(lua_State *L)
+{
+    /*
+      get the receive mutex for this driver
+     */
+    DroneCAN_Handle *h = check_DroneCAN_Handle(L, 1);
+    WITH_SEMAPHORE(h->dc->get_canard_iface().get_sem_rx());
+
+    auto *s = h->subscriber;
+
+    if (s == nullptr) {
+        return 0;
+    }
+    Subscriber::Payload payload;
+    if (!s->payloads.pop(payload)) {
+        return 0;
+    }
+
+    lua_pushlstring(L, (char *)payload.data, payload.length);
+    free(payload.data);
+
+    lua_pushinteger(L, payload.node_id);
+    *new_uint64_t(L) = payload.timestamp;
+    lua_pushboolean(L, payload.canfd);
+
+    if (s->trans_type == CanardTransferTypeResponse) {
+        // request reply removes the subscriber
+        delete h->subscriber;
+        h->subscriber = nullptr;
+    }
+    
+    return 4;
+}
+
+/*
+  cleanup any dangling subscribers
+ */
+void DroneCAN_Handle::destroy_all(void)
```

Comment:
This is not needed, Lua guarantees that the `__gc` of all objects will be called when the state is shut down, which we do properly.

Reply:
Even if something very bad happens? Maybe we could internal error if there is anything left for this to clean up and see if we get any reports.

Reply:
The behavior I mention is handled by `lua_close` which we always call AFAICT: "Destroys all objects in the given Lua state (calling the corresponding garbage-collection metamethods, if any) and frees all dynamic memory used by this state."

If there is memory corruption or something then the metamethod might not be called but then we are already hosed...

Reply:
I had tested this before with only the __gc and not the destroy_all() and I thought it did leak in some cases, but I now can't find a leak with either a script having an error or a scripting restart, so I've removed the destroy_all